### DISCUSSION TITLE
brew tries to install incorrect dependencies

### DISCUSSION THREAD

**Discussion by jacktose** [Homebrew/discussions#1210] [Linux]
When I `install` or `upgrade` various formulae, brew tries to install glibc, gmp, isl@0.18, mpfr, libmpc, and gcc@5 as dependencies, even though they are not dependencies for the formula in question.

[There's a full example in my gist][upgrade-gh], but here's the interesting part:
```
$ brew deps --include-build --include-optional --include-test --include-requirements gh
go
$ brew upgrade gh
==> Upgrading 1 outdated package:
gh 1.7.0 -> 1.8.1
==> Upgrading gh 1.7.0 -> 1.8.1 
... (warnings and downloads) ...
==> Installing dependencies for gh: glibc, gmp, isl@0.18, mpfr, libmpc and gcc@5
==> Installing gh dependency: glibc
^C
One sec, cleaning up...
```

This happens with various formulae, with different parameters. Here's vim, which does have other dependencies, and builds from source:
```
$ brew upgrade vim
==> Upgrading 1 outdated package:
vim 8.2.2725_1
==> Upgrading vim -> 8.2.2725_1 
...
==> Installing dependencies for vim: glibc, gmp, isl@0.18, mpfr, libmpc, gcc@5, lua, expat, sqlite, python@3.9 and ruby
==> Installing vim dependency: glibc
^C
One sec, cleaning up...
```

I'm aware that my setup is nonstandard ([see gist][gist]), which is why I didn't go straight to an issue for this, but I'd appreciate any help finding the cause and/or solution, and I'll happily submit a patch for it if appropriate.

[gist]: https://gist.github.com/jacktose/7207b73d376a5c6d7612ffeba6005324
[upgrade-gh]: https://gist.github.com/jacktose/7207b73d376a5c6d7612ffeba6005324#file-01_brew-upgrade-gh-txt

**Comment by alebcay** [comment]
My guess is that Homebrew deems your system's compiler setup to be too old or non-existent, which is why it's installing `gcc@5` (and dependencies of `gcc@5`: `gmp`, `isl@0.18`, `libmpc`, `mpfr`). I believe the `glibc` installation is also happening because your system's `glibc` is either too old or non-existent.

According to our docs [here](https://docs.brew.sh/Homebrew-on-Linux), your system _should_ be new enough that Homebrew should not be installing these things, but I would suspect that this is the general direction to investigate towards.

**Comment by Bo98** [comment]
> `/usr/bin/gcc: 4.8.5`

Homebrew for Linux requires GCC 5 to be installed (but is able to install its own if you have GCC 4.7.0 or later). This is a runtime dependency because of the linkage to `libstdc++` etc.

**Comment by jacktose** [comment]
Oh, that makes more sense. Before this, I uninstalled `gcc@5` etc. because I was trying to get the new, cleaner `gcc` setup to work.

So it's a legit installation, but a dependency of homebrew, not of the formula I'm installing directly. Seems like more of a UX issue then. Could it say something like
```
Homebrew requires `gcc@5` etc. on your system because your system gcc is <5.
```
or
```
==> Installing dependencies for Homebrew: glibc, gmp, isl@0.18, mpfr, libmpc and gcc@5
...
==> Installing dependencies for vim: lua, expat, sqlite, python@3.9 and ruby
```

**Reply by SMillerDev** [reply]
> Homebrew requires `gcc@5` etc. on your system because your system gcc is <5.

I believe it did this when you installed homebrew. If you uninstall it we don't keep track of the reasons anymore. We just know there is a dependency.
